From e8135b5e7b9720741289d59afb1b0d0d6a55e303 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Sun, 18 Jul 2021 20:16:42 +0300
Subject: [PATCH] [SW-50180] gaudi: fix pcie phy fw ext_ld_done loop

Commit
3c60806 ("[SW-13500] gaudi: fix instability in pci lanes")
while fixing the lanes instability introduced a logical bug - instead
of waiting for all PHYs to finish, checked that there is still a PHY
that has not finished.
So in case we've got that portion of code when all the PHYs have
finished it could result in an infinite loop - FW stuck.
Additionally the bug can result in link establishment problem as it
permits H/W bug W/A execution before PHYs have finished loading the FW.

Fixes: 3c6080670f4a314992c87481ddb051f1def33a09

Change-Id: Ia4edc1474b37186e913b262725a5b862ae6f3721
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/arch/gaudi/common.c b/arch/gaudi/common.c
index d87b281..3d38570 100644
--- a/arch/gaudi/common.c
+++ b/arch/gaudi/common.c
@@ -227,10 +227,10 @@
 
 	start = timestamp_get_us();
 	/* Wait for ext_ld_done before executing W/A */
-	while (regmap_readl(mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_0) &&
+	while (!(regmap_readl(mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_0) &&
 	       regmap_readl(mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_1) &&
 	       regmap_readl(mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_2) &&
-	       regmap_readl(mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_3)) {
+	       regmap_readl(mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_3))) {
 		/*
 		 * PCIe external load complete should take maximum of 300us
 		 * to complete. Including some grace period 576us of
